kentikapi/client_integration_test.go

```diff
@@ -426,83 +427,179 @@ func durationPtr(v time.Duration) *time.Duration {
+			s.Start()
+			defer s.Stop()
+
+			client, err := kentikapi.NewClient(kentikapi.Config{
+				SyntheticsGRPCHostPort: s.url,
+				AuthToken:              dummyAuthToken,
+				AuthEmail:              dummyAuthEmail,
+				DisableTLS:             true,
+			})
+			require.NoError(t, err)
+
+			var header metadata.MD
+			response, err := client.SyntheticsAdmin.GetAgent(
+				context.Background(),
+				tt.request,
+				grpc.Header(&header),
+			)
 
-			response, err := client.GRPCSyntheticsAdminServiceAPI.GetAgent(ctx, request)
+			if tt.expectedError {
+				assert.Error(t, err)
+			} else {
+				assert.NoError(t, err)
+			}
 
-			if response != nil {
-				assert.Equal(t, tt.res.GetAgent(), response.GetAgent())
+			if e, ok := status.FromError(err); ok {
+				assert.Equal(t, tt.expectedErrorCode, e.Code())
+				assert.Equal(t, tt.expectedErrorMsg, e.Message())
 			}
 
-			if err != nil {
-				assert.Errorf(t, err, tt.errMsg)
+			assert.Equal(t, len(s.responses), len(s.requests), "invalid number of requests")
+			for _, r := range s.requests {
+				assert.Equal(t, dummyAuthEmail, header.Get(authEmailKey)[0])
+				assert.Equal(t, dummyAuthToken, header.Get(authAPITokenKey)[0])
+				assert.Equal(t, tt.expectedRequestBody, r.String())
 			}
+
+			assert.Equal(t, tt.expectedResult.GetAgent(), response.GetAgent())
 		})
 	}
 }
 
-type mockSyntheticsServer struct {
-	pb.UnimplementedSyntheticsAdminServiceServer
+type spySyntheticsServer struct {
+	syntheticspb.UnimplementedSyntheticsAdminServiceServer
+	server    *grpc.Server
+	url       string
+	t         testing.TB
+	requests  []*syntheticspb.GetAgentRequest
+	responses []*gRPCResponse
+}
+
+type gRPCResponse struct {
+	err  error
+	body *syntheticspb.GetAgentResponse
+}
+
+func newSpySyntheticsServer(t testing.TB, responses []*gRPCResponse) *spySyntheticsServer {
+	return &spySyntheticsServer{
+		t:         t,
+		responses: responses,
+	}
+}
+
+func (s *spySyntheticsServer) Start() {
+	l, err := net.Listen("tcp", "localhost:0")
+	require.NoError(s.t, err)
+
+	s.url = l.Addr().String()
+
+	s.server = grpc.NewServer()
+	syntheticspb.RegisterSyntheticsAdminServiceServer(s.server, s)
+	go func() {
+		require.NoError(s.t, s.server.Serve(l))
+	}()
+}
+
+func (s *spySyntheticsServer) Stop() {
+	s.server.GracefulStop()
+}
+
+func (s *spySyntheticsServer) GetAgent(ctx context.Context, req *syntheticspb.GetAgentRequest,
+) (*syntheticspb.GetAgentResponse, error) {
+	header, ok := metadata.FromIncomingContext(ctx)
+	assert.True(s.t, ok)
+	assert.NoError(s.t, grpc.SendHeader(ctx, header))
+
+	s.requests = append(s.requests, req)
+
+	response, err := s.response()
+	assert.NoError(s.t, err)
+
+	return response.body, response.err
 }
 
-func (*mockSyntheticsServer) GetAgent(_ context.Context, req *pb.GetAgentRequest) (*pb.GetAgentResponse, error) {
-	if req.Id == "968" {
-		return &pb.GetAgentResponse{Agent: newDummyAgentForGRPC()}, nil
+func (s *spySyntheticsServer) response() (*gRPCResponse, error) {
```

Comment:
Could you please remove error from this function signature and return gRPCResponse with error field set. Currently returning nil grpcresponse will lead to handler panic, which we want to avoid.

Reply:
Done